File: muesli/packages/MuesliCore/Sources/MuesliCore/Transcript.swift
import Foundation

public struct Utterance: Codable, Equatable {
    /// Stable index; notes.md spans reference these. Strictly increasing.
    public let i: Int
    public let t0: Double
    public let t1: Double
    public let speaker: String?
    public let text: String
    public let conf: Double?

    public init(i: Int, t0: Double, t1: Double, speaker: String? = nil, text: String, conf: Double? = nil) {
        self.i = i
        self.t0 = t0
        self.t1 = t1
        self.speaker = speaker
        self.text = text
        self.conf = conf
    }
}

public enum Transcript {
    public static func parse(_ jsonl: String) throws -> [Utterance] {
        let decoder = JSONDecoder()
        var utterances: [Utterance] = []
        let lines = jsonl.split(whereSeparator: \.isNewline).filter {
            !$0.trimmingCharacters(in: .whitespaces).isEmpty
        }
        for (index, line) in lines.enumerated() {
            guard let data = String(line).data(using: .utf8),
                let u = try? decoder.decode(Utterance.self, from: data)
            else {
                throw FormatError.transcript("line \(index + 1): invalid utterance")
            }
            utterances.append(u)
        }
        for n in 1..<max(utterances.count, 1) {
            if utterances[n].i <= utterances[n - 1].i {
                throw FormatError.transcript("utterance index not strictly increasing at line \(n + 1)")
            }
            if utterances[n].t0 < utterances[n - 1].t0 {
                throw FormatError.transcript("utterances not ordered by t0 at line \(n + 1)")
            }
        }
        return utterances
    }

    public static func serialize(_ utterances: [Utterance]) throws -> String {
        var lines: [String] = []
        for u in utterances {
            // Hand-assemble the object so key order is stable across platforms
            // (mirrors the TypeScript writer: i, t0, t1, speaker?, text, conf?).
            var fields: [String] = []
            fields.append("\"i\": \(u.i)")
            fields.append("\"t0\": \(formatNumber(u.t0))")
            fields.append("\"t1\": \(formatNumber(u.t1))")
            if let speaker = u.speaker {
                fields.append("\"speaker\": \(try encodeString(speaker))")
            }
            fields.append("\"text\": \(try encodeString(u.text))")
            if let conf = u.conf {
                fields.append("\"conf\": \(formatNumber(conf))")
            }
            lines.append("{" + fields.joined(separator: ", ") + "}")
        }
        return lines.joined(separator: "\n") + "\n"
    }

    private static func formatNumber(_ n: Double) -> String {
        if n == n.rounded(), n.magnitude < 1e15 { return String(Int64(n)) }
        return String(n)
    }

    private static func encodeString(_ s: String) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.withoutEscapingSlashes]
        return String(data: try encoder.encode(s), encoding: .utf8)!
    }

    /// Render the human-readable transcript.md from utterances + speaker names.
    public static func renderMarkdown(
        title: String, utterances: [Utterance], speakerNames: [String: String] = [:]
    ) -> String {
        let lines = utterances.map { u -> String in
            let mm = Int(u.t0) / 60
            let ss = Int(u.t0) % 60
            let stamp = String(format: "%02d:%02d", mm, ss)
            var label = ""
            if let speaker = u.speaker {
                label = " \(speakerNames[speaker] ?? speaker):"
            }
            return "**[\(stamp)]\(label)** \(u.text)"
        }
        return "# Transcript — \(title)\n\n" + lines.joined(separator: "\n\n") + "\n"
    }
}

File: muesli/packages/muesli-format/src/transcript.ts
export interface Utterance {
  /** Stable index; notes.md spans reference these. Strictly increasing. */
  i: number;
  t0: number;
  t1: number;
  speaker?: string;
  text: string;
  conf?: number;
}

export function parseTranscript(jsonl: string): Utterance[] {
  const utterances: Utterance[] = [];
  const lines = jsonl.split(/\r?\n/).filter((l) => l.trim() !== "");
  for (const [lineNo, line] of lines.entries()) {
    let obj: unknown;
    try {
      obj = JSON.parse(line);
    } catch {
      throw new Error(`transcript line ${lineNo + 1}: invalid JSON`);
    }
    const u = obj as Record<string, unknown>;
    if (
      typeof u.i !== "number" ||
      typeof u.t0 !== "number" ||
      typeof u.t1 !== "number" ||
      typeof u.text !== "string"
    ) {
      throw new Error(`transcript line ${lineNo + 1}: missing required fields`);
    }
    utterances.push(u as unknown as Utterance);
  }
  for (let n = 1; n < utterances.length; n++) {
    if (utterances[n].i <= utterances[n - 1].i) {
      throw new Error(`transcript: utterance index not strictly increasing at line ${n + 1}`);
    }
    if (utterances[n].t0 < utterances[n - 1].t0) {
      throw new Error(`transcript: utterances not ordered by t0 at line ${n + 1}`);
    }
  }
  return utterances;
}

export function serializeTranscript(utterances: Utterance[]): string {
  return (
    utterances
      .map((u) => {
        const ordered: Record<string, unknown> = { i: u.i, t0: u.t0, t1: u.t1 };
        if (u.speaker !== undefined) ordered.speaker = u.speaker;
        ordered.text = u.text;
        if (u.conf !== undefined) ordered.conf = u.conf;
        return JSON.stringify(ordered);
      })
      .join("\n") + "\n"
  );
}

/** Render the human-readable transcript.md from utterances + speaker names. */
export function renderTranscriptMarkdown(
  title: string,
  utterances: Utterance[],
  speakerNames: Record<string, string> = {},
): string {
  const lines = utterances.map((u) => {
    const mm = Math.floor(u.t0 / 60);
    const ss = Math.floor(u.t0 % 60);
    const stamp = `${String(mm).padStart(2, "0")}:${String(ss).padStart(2, "0")}`;
    const who = u.speaker ? (speakerNames[u.speaker] ?? u.speaker) + ":" : "";
    const label = who ? ` ${who}` : "";
    return `**[${stamp}]${label}** ${u.text}`;
  });
  return `# Transcript — ${title}\n\n${lines.join("\n\n")}\n`;
}
